### DISCUSSION TITLE
urequests formatting & urlencoding

### DISCUSSION THREAD

**Discussion by theseant** [micropython/micropython#11311] [RP2040 / Pico]
I have functioning python code to request data from the husqvarna api. I have tried to recreate this fucntionality in micropython on the Pi Pico with a view to making a display that gives the current status of a husqvarna automower. However, I cannot work out how to format the urequests.post.

I found a really useful topic in the archive (https://forum.micropython.org/viewtopic.php?t=8802) and using the advice from @Jimmo, I seem to have almpost solved the problem. The request is sent but the api returns an error
``` 
("error":"invalid_client","error_description":"Invalid` client: cannot retrieve client `credentials","error_code":"client.not.found").
```
There still seems to be an error in the formatting so that the api does not recognise the client_details. 

My micropython code:
```
import network
import time
import urequests
from connect import connect
import secrets
import ujson

api_url = "https://api.authentication.husqvarnagroup.dev/v1/oauth2/token"

def url_escape(s):
    return ''.join(c if c.isalpha() or c.isdigit() else '%%%02x' % ord(c) for c in s)

#use a prewritten fucntion to connect to wifi to save rewriting
connect()

parameters = ({
    'grant_type': 'client_credentials',
    'client_id': secrets.husq_client_ID,
    'client_secret': secrets.husq_client_secret
})
print("Attempting to send Prompt")
r = urequests.post(api_url, data=url_escape(ujson.dumps(parameters)), headers={'Content-Type': 'application/x-www-form-urlencoded'})
 
if r.status_code >= 300 or r.status_code < 200:
    print("There was an error with your request \n" +
              "Response Status: " + str(r.text))
else:
    print("Success")
    response_data = json.loads(r.text)
    completion = response_data
    print(completion)
    r.close()  
```
The console output is 
```
Attempting to send Prompt
There was an error with your request 
Response Status: {"error":"invalid_client","error_description":"Invalid client: cannot retrieve client credentials","error_code":"client.not.found"}  
```
Printing data=url_escape(ujson.dumps(parameters)) to the consolegives (scerets obscured):
```
%7b%22grant%5ftype%22%3a%20%22client%5fcredentials%22%2c%20%22client%5fsecret%22%3a%20%22*******%2d****%2d****%2d****%2d************%22%2c%20%22client%5fid%22%3a%20%22********%2d****%2d****%2d***%2************%22%7d
```
Any advice would be greatly appreciated.

**Comment by karfas** [comment]
Maybe it's better to send the JSON as `application/json` and don't encode the JSON ?

**Reply by theseant** [reply]
Thank you. I have tried this using the code from https://makeblock-micropython-api.readthedocs.io/en/latest/public_library/Third-party-libraries/urequests.html.

I may not have deone this correctly but it does send and I get a response - unfortunately, not the response I want.

The new code reads
```
post_data = ujson.dumps({'grant_type': 'client_credentials', 'client_id': secrets.husq_client_ID, 'client_secret': secrets.husq_client_secret})
print("Attempting to send Prompt")

r = urequests.post(api_url, headers = {'content-type': 'application/json'}, data = post_data)
if r.status_code >= 300 or r.status_code < 200:
    print("There was an error with your request \n" +
              "Response Status: " + str(r.text))
```
and the response seems to make it clear it shoudl url encoded?
```
Response Status: {"error":"invalid_request","error_description":"Invalid request: content must be application/x-www-form-urlencoded","error_code":"invalid.request"}
```

**Reply by karfas** [reply]
I got fooled by your usage of JSON in the first post.

Just find it out yourself. Husqvarna has a nice instance of an OpenAPI client at https://developer.husqvarnagroup.cloud/apis/authentication-api#openapi - click on one of the `Post` buttons and `try it out`, fill in your credentials and look at the request + response.

Note that a www-form-urlencoded has nothing to do with JSON (check wikipedia if you are unsure).

**Reply by theseant** [reply]
Thanks. I have used the Husqvarna API client. Although there aren't any specific python resources, I have managed to get the API working in Python 3 on a raspberry pi using code below.
```
api_parameters = {
    "grant_type": grant_type,
    "client_id": client_ID,
    "client_secret": client_secret
}

response = requests.post(api_url, api_parameters)
completion = json.loads(response.text)
token = completion["access_token"]

mowers_api_url = 'https://api.amc.husqvarna.dev/v1/mowers'
headers = {
    'Authorization':'Bearer '+token,
    'X-Api-Key': secrets.husq_client_ID,
    'Authorization-Provider': 'husqvarna'
}
mowers = requests.get(url=mowers_api_url, headers=headers)
mowersdetails = json.loads(mowers.text)
```
I just can't get the syntax correct for urequests and micropython so that I can run it on a Pi Pico.

**Comment by andypiper** [comment]
Is your Python code sending the client credentials via JSON or parameters in the post body, or via the HTTP headers?

**Reply by theseant** [reply]
Due to my limited python and api knowledge, I am not sure of the answer, But in the GET request example, I think it is in parameters in the post body based on the working python code. 

In the second part (using the authorisation token to access specific details), it would appear to be in the headers of the .POST request.
